src/wfs/src/test/java/org/geoserver/wfs/TransactionListenerTest.java

```diff
@@ -131,6 +131,48 @@ public void testInsert() throws Exception {
         assertEquals(fid, inserted.getIdentifier().getID());
     }
 
+    @Test
+    public void testOrderedInsert() throws Exception {
+        // perform an insert
+        String insert = "<wfs:Transaction service=\"WFS\" version=\"1.0.0\" "
+                + "xmlns:cgf=\"http://www.opengis.net/cite/geometry\" "
+                + "xmlns:ogc=\"http://www.opengis.net/ogc\" "
+                + "xmlns:wfs=\"http://www.opengis.net/wfs\" "
+                + "xmlns:gml=\"http://www.opengis.net/gml\"> "
+                + "<wfs:Insert > "
+                + "<cgf:Lines>"
+                + "<cgf:lineStringProperty>"
+                + "<gml:LineString>"
+                + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056415,5433016.8189323 494982.70115662,5435041.95096618"
+                + "</gml:coordinates>"
+                + "</gml:LineString>"
+                + "</cgf:lineStringProperty>"
+                + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>"
+                + "<cgf:Points>"
+                + "<cgf:pointProperty>"
+                + "<gml:Point><gml:coordinates>15,15</gml:coordinates></gml:Point>"
+                + "</cgf:pointProperty>"
+                + "<cgf:id>t0005</cgf:id>"
+                + "</cgf:Points>"
+                + "</wfs:Insert>"
+                + "</wfs:Transaction>";
+
+        postAsDOM("wfs", insert);
+        assertEquals(4, listener.features.size());
+
+        // get the feature before insert
```

Comment:
The insert only has two features, with id t0002 and t0005, however they are found 4 times because of pre and post-insert events (2 events for each feature). It would be clearer if you tested the events and the features inside of them.